src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -0,0 +1,749 @@
+     *
+     * @return a {@code Future} that will succeed when the client has connected to the gateway and is ready to send
+     * push notifications
+     *
+     * @see ApnsClient#PRODUCTION_APNS_HOST
+     * @see ApnsClient#DEVELOPMENT_APNS_HOST
+     * @see ApnsClient#DEFAULT_APNS_PORT
+     * @see ApnsClient#ALTERNATE_APNS_PORT
+     */
+    public Future<Void> connect(final String host, final int port) {
+        final Future<Void> connectionReadyFuture;
+
+        if (this.bootstrap.group().isShuttingDown() || this.bootstrap.group().isShutdown()) {
+            connectionReadyFuture = new FailedFuture<Void>(GlobalEventExecutor.INSTANCE,
+                    new IllegalStateException("Client's event loop group has been shut down and cannot be restarted."));
+        } else {
+            synchronized (this.bootstrap) {
+                // We only want to begin a connection attempt if one is not already in progress or complete; if we already
+                // have a connection future, just return the existing promise.
+                if (this.connectionReadyPromise == null) {
+                    final ChannelFuture connectFuture = this.bootstrap.connect(host, port);
+                    this.connectionReadyPromise = connectFuture.channel().newPromise();
+
+                    connectFuture.addListener(new GenericFutureListener<ChannelFuture>() {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            if (!future.isSuccess()) {
+                                log.debug("Failed to connect.", future.cause());
+                                ApnsClient.this.connectionReadyPromise.tryFailure(future.cause());
+                            }
+                        }
+                    });
+
+                    connectFuture.channel().closeFuture().addListener(new GenericFutureListener<ChannelFuture> () {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            // We always want to try to fail the "connection ready" promise if the connection closes; if
+                            // it has already succeeded, this will have no effect.
+                            ApnsClient.this.connectionReadyPromise.tryFailure(
+                                    new IllegalStateException("Channel closed before HTTP/2 preface completed."));
+
+                            synchronized (ApnsClient.this.bootstrap) {
+                                ApnsClient.this.connectionReadyPromise = null;
+
+                                if (ApnsClient.this.reconnectionPromise != null) {
+                                    log.debug("Disconnected. Next automatic reconnection attempt in {} seconds.", ApnsClient.this.reconnectDelay);
+
+                                    future.channel().eventLoop().schedule(new Runnable() {
```

Comment:
That's definitely true, but it seems like there's some badness either way. Whether or not we stop the automatic reconnection attempts, callers are left in a situation where their client has stopped and will not reconnect. As near as I can tell, most of what we'd gain by stopping reconnection attempts after certificate expiration is a modest reduction in resource usage (which should already be very low thanks to exponential back-off) until somebody intervenes.

It's true that we could offer users a clear warning about expiration if we think it might be a problem, but I suspect that will also be clear from the exceptions generated when reconnection attempts fail. Given that we could also conceivably get things wrong with this kind of check (and accidentally clobber clients that would otherwise reconnect), I'm inclined to leave certificate expiration handling out of Pushy itself and trust other layers of the whole system to deal with it.

Seem okay?